docs/dsl/example.md

```diff
@@ -0,0 +1,438 @@
+| 51 | 4567 | flu | 2021-12-14 |
+| 52 | 3456 | pf | 2021-09-09 |
+| 53 | 2345 | flu | 2021-11-02 |
+| 54 | 5678 | pf | 2021-07-01 |
+| 55 | 3456 | pox | 2021-11-03 |
+| 56 | 3456 | az | 2021-04-01 |
+| 57 | 5678 | az | 2021-12-30 |
+| 58 | 5678 | pf | 2021-07-02 |
+
+**patients** is a `PatientFrame`, while **registrations** and **immunisations** are `EventFrame`s.
+
+---
+
+A `PatientFrame` represents a collection of rows with at most one row per patient.  An `EventFrame` represents as collection of rows with multiple rows per patient.
+
+An `EventFrame`/`PatientFrame` can be filtered to produce another `EventFrame`/`PatientFrame`, and you should think of them as being like unevaluated Django Querysets, in that they can be combined and acted on to produce something that can be turned into SQL.
+
+Ultimately, we transform an `EventFrame`/`PatientFrame` into a `PatientSeries`, through filtering, sorting, aggregating, and selecting a column.
+
+A `PatientSeries` represents a collection of values with one value per patient.
+
+A `PatientSeries` can be assigned to a cohort, in which case a column with the `PatientSeries`'s values will be appended to the cohort's input.csv.
+
+---
+
+This diagram attempts to explain how a `EventFrame`/`PatientFrame` is transformed into a
+`PatientSeries`.
+
+The nodes are Python classes and the arrows are methods on those classes.
+
+```
+             +---+
+             |   V
+      filter |  PatientFrame
+             |   |   |
+             +---+   | select_column
+                     V
+                PatientSeries
+
+
+             +---+
+             |   V
+      filter |  EventFrame -----------------------+
+             |   |   |                            |
+             +---+   | sort_by                    |
+                     V                            |
+             SortedEventFrame                     |
+                     |                            |
+        +---+        | (first/last)_for_patient   | (count/exists)_for_patient
+        |   V        V                            |
+ filter |  AggregatedEventFrame                   |
+        |   |        |                            |
+        +---+        | select_column              |
+                     V                            |
+               PatientSeries <--------------------+
+```
+
+---
+
+Here's the code that we can use to generate an input.csv table.
+
+```
+from cohortextractor2.concepts.tables import (
+    immunisations as imms,
+    patients,
+    registrations,
+)
+from cohortextractor2.dsl import Cohort, categorise
+
+from codelists import covid_vacc_codes
+
+# Create a cohort.
+cohort = Cohort()
+
+# Create two PatientSeries for age and sex and assign to the cohort.
+cohort.age = patients.select_column(patients.age)
+cohort.sex = patients.select_column(patients.sex)
+
+# Create a new EventFrame representing just the events with covid vaccination codes.
+covid_imms = imms.filter(covid_vacc_codes.contains(imms.code))
```

Comment:
thought: What I can do with `filter` and what I can pass to it, isn't made explicit in this document?